### ISSUE TITLE
[TODO] [Day 5] CNN / Transformer 백본 모델 이해 및 efficient net 모델 개선 

### ISSUE THREAD

**Issue by nimnusmik** [AIBootcamp14/computervisioncompetition-cv-1#32]
## 📋 작업 개요
**작업 주제:** 문서 분류 모델 성능 최적화 - 단계별 개선 실험  
**담당자:** @kimsunmin0227  
**마감일:** 9/6 24:00  
**예상 소요시간:** 총 8시간 (실험 4시간 + 분석/문서화 4시간)

## 📅 기간
- 시작일: 2025-09-04
- 종료일: 2025-09-06

## 📌 목표
- 현재 CV F1 0.928에서 0.95+ 달성
- 각 개선사항의 개별 성능 기여도 측정
- 재현 가능한 최적화 파이프라인 구축
- WandB를 통한 체계적인 실험 관리

### Phase 1: 모델 아키텍처 개선  
- [ ] Multi-scale Feature Extraction 구현
- [ ] Attention Mechanism 추가 (Channel + Spatial)
- [ ] GeM Pooling 적용
- [ ] 개선된 Classifier Head 구성
- [ ] 성능 비교 실험 (baseline vs improved)

### Phase 2: 고급 데이터 증강 
- [x] 적응적 증강 전략 구현
- [x] CutMix 추가 구현
- [ ] 문서 특화 Perspective/Elastic 변환 최적화
- [ ] 증강 강도 스케줄링 적용

### Phase 3: 손실 함수 및 정규화 개선  
- [ ] Focal Loss 구현 및 적용
- [ ] Class-balanced weights 계산
- [ ] SAM (Sharpness-Aware Minimization) 도입
- [ ] Label Smoothing 최적화

### Phase 4: 학습 전략 최적화  
- [ ] OneCycleLR + Warmup 적용
- [ ] Gradient Checkpointing 구현
- [ ] 동적 배치 크기 조정
- [ ] Early Stopping 전략 개선

### Phase 5: 앙상블 다양성 증대 
- [ ] 다양한 아키텍처 조합 (EfficientNet + ConvNeXt + Swin)
- [ ] 가중 앙상블 최적화
- [ ] TTA 전략 고도화
- [ ] Temperature Scaling 개선

## 💡 참고 사항
- 각 단계별로 개별 실험 후 성능 측정 필수
- WandB 대시보드에서 실시간 모니터링
- 메모리 사용량 지속적 체크 (GPU 16GB 한계)
- 기존 코드 백업 후 점진적 개선

### 유용한 링크
- [x] Kaggle 유사 솔루션: [[Document Classification Winners](https://www.kaggle.com/competitions/benetech-making-graphs-accessible/discussion)](https://www.kaggle.com/competitions/benetech-making-graphs-accessible/discussion)
- [x] PyTorch 공식 예제: [[Image Classification Tutorial](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html)](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html)
- [x] Timm 라이브러리 문서: [[Timm Documentation](https://timm.fast.ai/)](https://timm.fast.ai/)
- [ ] SAM Optimizer: [[GitHub Repository](https://github.com/davda54/sam)](https://github.com/davda54/sam)
- [ ] CutMix 구현: [[Official Paper Implementation](https://github.com/clovaai/CutMix-PyTorch)](https://github.com/clovaai/CutMix-PyTorch)

---

### 실험 중 발견사항
- 현재 모델이 Class 1(46개), Class 14(50개) 등 소수 클래스에서 성능 저하
- TTA 시 신뢰도가 평균 0.416으로 낮음 → 모델 확신도 개선 필요
- Early Stopping이 모든 Fold에서 발생 → 더 긴 학습 필요할 수 있음

### 다음 실험 아이디어
- Knowledge Distillation으로 큰 모델→작은 모델 전이
- AutoAugment 정책 학습
- Multi-task Learning (회전 각도 예측 보조 태스크)

### 팀원 피드백
- 메모리 효율성 우선 고려 (16GB GPU 한계)
- 재현성을 위한 시드 고정 철저히
- 각 개선사항별 개별 성능 기여도 정량화 필요
- 최종 앙상블에서 추론 시간도 고려해야 함

**⚠️ 주의사항:** 각 Phase 완료 후 반드시 성능 검증하고 다음 단계 진행. 성능 저하 시 롤백 전략 준비.

**Comment by nimnusmik** [comment]
# 클래스 혼동 개선 전략 (우선순위별)

## 📋 개요
클래스 3, 7, 14 간 혼동을 줄이는 실전 전략들을 OCR 없이 해결하는 방법들로 정리

---

## 🏆 1순위: 간단하지만 효과 빠른 전처리·입력 변형

### 1.1 문서 전용 전처리 채널 추가
- **방법**: RGB + Binary + Edge를 3채널로 합쳐 모델 입력으로 사용
- **파라미터**: 
  - Adaptive threshold: `blockSize=15~51, C=8`
  - Canny edge: `thresholds=(50,150)`
- **기대효과**: 텍스트·레이아웃 대비 노이즈(조명, 색상) 영향 약화

### 1.2 CLAHE 적용
- **방법**: 밝기·대비 불균일한 문서 사진에 적용
- **파라미터**: `clipLimit=2.0, tileGridSize=(8,8)`
- **기대효과**: 글자·세부 요소 강조로 시각적 차이 증폭

### 1.3 문서 중심 크롭/기울기 보정
- **방법**: 최대 연결 영역 찾아 문서 중심으로 crop & deskew
- **기대효과**: 배경/테두리 변화 제거로 문서 내부 패턴에 집중

---

## 🥈 2순위: 데이터 증강/샘플링의 정교화

### 2.1 문서 특화 augmentation
- **방법**: 
  - Perspective warp (약한 정도)
  - Small rotations ±3-8°
  - Random brightness/contrast ±10-20%
  - Gaussian blur sigma 0~1
- **기대효과**: 촬영 각도·부분 가림 상황에 강해짐

### 2.2 Hard Example Synthesis
- **방법**: 헷갈리는 클래스(3,7,14) 내에서 강한 변형 생성
- **기대효과**: intra-class 다양성 증가로 결정 경계 선명화

### 2.3 Class-balanced Oversampling
- **방법**: 검증에서 많이 틀리는 클래스를 훈련에서 더 자주 노출
- **기대효과**: 어려운 클래스에 대한 학습 강화

---

## 🥉 3순위: 모델/학습 기법

### 3.1 Metric Learning
- **방법**: Cross-entropy + Triplet loss 또는 SupCon 병행
- **파라미터**: `margin=0.2, temp=0.07`
- **기대효과**: 클래스 간 임베딩 거리 확대로 혼동 감소

### 3.2 Center Loss / ArcFace
- **방법**: CrossEntropy + CenterLoss 또는 ArcFace 적용
- **기대효과**: 같은 클래스 응집, 다른 클래스와 분리

### 3.3 Multi-head / Multi-task
- **방법**: 레이아웃·문서유형·스탬프 여부 등 보조 태스크 추가
- **기대효과**: 문서 구조를 더 잘 포착

### 3.4 Backbone 다양화
- **방법**: ConvNeXt-Tiny, EfficientNet 혼합 앙상블
- **기대효과**: 서로 다른 inductive bias로 보완

---

## 🛠️ 4순위: 구조적 접근

### 4.1 Patch-based Aggregation
- **방법**: 여러 패치(224×224) 추출 → 각 패치 분류 → soft voting
- **기대효과**: 로컬 특징 학습 강화

### 4.2 Vision Transformer
- **방법**: ViT/DeiT/Swin으로 교체
- **기대효과**: 문서 레이아웃·구성 차이를 더 잘 포착

### 4.3 OCR-free 문서 모델
- **방법**: Donut/DocFormer 같은 전용 모델
- **기대효과**: OCR 없이도 문서 타입·키워드 파악

---

## 🔍 5순위: 반자동 데이터 작업

### 5.1 Hard Negative Mining
- **방법**: 모델이 틀리거나 낮은 신뢰도 샘플로 별도 fine-tune
- **기대효과**: 경계 샘플에 대한 결정력 개선

### 5.2 Self-supervised Pretraining
- **방법**: DINO, MoCo로 도메인 특화 표현 학습 후 fine-tune
- **기대효과**: 문서 특유 패턴을 더 잘 포착

---

## 🧪 권장 실험 순서

1. **Grad-CAM 확인** - 모델이 어디를 보는지 진단
2. **입력 채널 실험** - RGB vs (RGB+binary+edge) 비교
3. **CLAHE + deskew** 파이프라인 적용 후 재훈련
4. **Patch-based voting** 간단 구현으로 검증 비교
5. **Metric learning** 추가로 임베딩 품질 향상
6. **Weight decay** 추가 (현재 없음 → 1e-4 적용)
7. **MixUp/CutMix ablation** - 문서 클래스용 조정

## 💡 구체적 설정 예시

### 전처리 파이프라인
```python
# 1. 원본 RGB 리사이즈 (비율 보존)
# 2. 그레이스케일 → CLAHE(clip=2.0) → adaptiveThreshold(block=35,C=8) → Canny(50,150)
# 3. [CLAHE, adaptive, Canny] 3채널로 합성
```

### Triplet Loss 설정
```python
margin = 0.2
backbone_output = 512  # dimension
lambda_triplet = 0.5
total_loss = CE_loss + lambda_triplet * triplet_loss
```

### AdamW 설정
```python
lr = 5e-4  # 기존 유지
weight_decay = 1e-4  # 새로 추가
betas = (0.9, 0.999)
```

각 단계마다 Confusion Matrix(특히 클래스 3, 7, 14의 precision/recall)로 성능 비교 필수!
